source/pdu/auto_switch/auto_switch.c

```diff
@@ -0,0 +1,182 @@
+    auto_switch.current[SW_MAIN] = calcCurrent_LP(adc_readings.main_cs);
+    // TODO: bullet switch doesn't have cs signal
+
+    // Upstream CS
+    calcCurrent_Total();
+}
+
+// Called periodically, Updates voltage for each rail in mV
+void updateVoltage() {
+    auto_switch.voltage.in_24v = calcVoltage(adc_readings.lv_24_v_sense, LV_24V_R1, LV_24V_R2);
+    auto_switch.voltage.out_5v = calcVoltage(adc_readings.lv_5_v_sense, LV_5V_R1, LV_5V_R2);
+    auto_switch.voltage.out_3v3 = calcVoltage(adc_readings.lv_3v3_v_sense, LV_3V3_R1, LV_3V3_R2);
+}
+
+// Current helper functions
+uint16_t calcCurrent_HP(uint16_t current) {
+    current = current * ADC_REF_mV / ADC_MAX;  // Convert to mV
+    current = current * (HP_CS_R1 + HP_CS_R2) / HP_CS_R2;
+    current = current * HP_CS_R3 / (HP_CS_R1 + HP_CS_R2);
+    return current;
+}
+
+uint16_t calcCurrent_LP(uint16_t current) {
+    current = current * ADC_REF_mV / ADC_MAX;  // Convert to mA
+    return current;
+}
+
+// CS signals for upstream 24V and 5V (total current through each)
+void calcCurrent_Total() {
+    // 24V current
+    uint16_t current = adc_readings.lv_24_i_sense;
+    current = current * ADC_REF_mV / ADC_MAX;  // Convert to mV
+    current = current / HP_CS_R_SENSE / CS_GAIN;
+    auto_switch.current[CS_24V] = current;
+
+    // 5V current
+    current = adc_readings.lv_5_i_sense;
+    current = current * ADC_REF_mV / ADC_MAX;  // Convert to mA
+    auto_switch.current[CS_5V] = current;
+}
+
+// Converts ADC voltage reading to mV
+uint16_t calcVoltage(uint16_t voltage, int r1, int r2) {
+    // Compensate for voltage divider
+    voltage = voltage * (r1 + r2) / r2;
+    // Scale for 12-bit adc at 3.3v
+    voltage = voltage * ADC_REF_mV / ADC_MAX;
+    return voltage;
+}
+
+// Enable or disable switche by name
+void enableSwitch(uint16_t auto_switch_enum, uint16_t value) {
+    switch (auto_switch_enum) {
+        case SW_PUMP_1:
+            PHAL_writeGPIO(PUMP_1_CTRL_GPIO_Port, PUMP_1_CTRL_Pin, value);
+            break;
+        case SW_PUMP_2:
+            PHAL_writeGPIO(PUMP_2_CTRL_GPIO_Port, PUMP_2_CTRL_Pin, value);
+            break;
+        case SW_SDC:
+            PHAL_writeGPIO(SDC_CTRL_GPIO_Port, SDC_CTRL_Pin, value);
+            break;
+        case SW_AUX:
+            PHAL_writeGPIO(AUX_HP_CTRL_GPIO_Port, AUX_HP_CTRL_Pin, value);
+            break;
+        case SW_FAN_1:
+            PHAL_writeGPIO(FAN_1_CTRL_GPIO_Port, FAN_1_CTRL_Pin, value);
+            break;
+        case SW_FAN_2:
+            PHAL_writeGPIO(FAN_2_CTRL_GPIO_Port, FAN_2_CTRL_Pin, value);
+            break;
+        case SW_DASH:
+            PHAL_writeGPIO(DASH_CTRL_GPIO_Port, DASH_CTRL_Pin, value);
+            break;
+        case SW_ABOX:
+            PHAL_writeGPIO(ABOX_CTRL_GPIO_Port, ABOX_CTRL_Pin, value);
+            break;
+        case SW_MAIN:
+            PHAL_writeGPIO(MAIN_CTRL_GPIO_Port, MAIN_CTRL_Pin, value);
```

Comment:
Just an FYI, we do not want the ability to control Dash, A-Box, or Main as they are essential components to the vehicle and we should not be able to disable them (The fact we can is actually a rules violation). We have disabled the pin inits for these pins, but we should not allow any users of this library to control these three as that could compromise the legality/safety of the car.